File: python/sorting/counting_sort_production.py
from collections.abc import Iterable


def counting_sort(values: Iterable[int], *, max_range: int = 1_000_000) -> list[int]:
    items = list(values)

    if max_range <= 0:
        raise ValueError("max_range must be greater than zero")

    if not items:
        return []

    if any(not isinstance(value, int) for value in items):
        raise TypeError("counting_sort only accepts integers")

    minimum = min(items)
    maximum = max(items)
    value_range = maximum - minimum + 1

    if value_range > max_range:
        return sorted(items)

    counts = [0] * value_range

    for value in items:
        counts[value - minimum] += 1

    result: list[int] = []

    for offset, frequency in enumerate(counts):
        result.extend([minimum + offset] * frequency)

    return result

File: java/sorting/CountingSortProduction.java
import java.util.Arrays;
import java.util.Objects;

public final class CountingSortProduction {

    private static final int DEFAULT_MAX_RANGE = 1_000_000;

    private CountingSortProduction() {
    }

    public static int[] countingSort(int[] values) {
        return countingSort(values, DEFAULT_MAX_RANGE);
    }

    public static int[] countingSort(int[] values, int maxRange) {
        Objects.requireNonNull(values, "values");

        if (maxRange <= 0) {
            throw new IllegalArgumentException("maxRange must be greater than zero");
        }

        if (values.length == 0) {
            return new int[0];
        }

        int minimum = values[0];
        int maximum = values[0];

        for (int value : values) {
            if (value < minimum) {
                minimum = value;
            }

            if (value > maximum) {
                maximum = value;
            }
        }

        long valueRange = (long)maximum - minimum + 1L;

        if (valueRange > maxRange) {
            int[] result = Arrays.copyOf(values, values.length);
            Arrays.sort(result);
            return result;
        }

        int[] counts = new int[(int)valueRange];

        for (int value : values) {
            counts[value - minimum]++;
        }

        int[] result = new int[values.length];
        int position = 0;

        for (int offset = 0; offset < counts.length; offset++) {
            int value = minimum + offset;

            for (int count = 0; count < counts[offset]; count++) {
                result[position++] = value;
            }
        }

        return result;
    }
}
